Driver3SchemaBuilder rework

avelanarius opened the pull request:
This PR is a bulk of work need for proper schema change support in the library.

The first commit renames `ColumnType` to `ColumnKind`. `ColumnType` is not removed, but deprecated (and similarly the getter). Added missing JavaDocs. 

The second commit ("Rework of Driver3SchemaBuilder") is essentially a complete rework of `Driver3SchemaBuilder` (therefore I suggest not looking at a diff, but a final file). The new implementation of `Driver3SchemaBuilder` relies on both CDC log schema and base table schema, but it uses the CDC log as much as possible. For certain information, it must go to the base table schema, which could be stale. `BaseTableSchema` class is introduced for this purpose and all operations are "proven" to work correctly (see tables in JavaDoc), even under possible inconsistencies (or it throws `UnresolvableSchemaInconsistencyException`). The generated `ChangeSchema` is consistent with the CDC log schema.

The third and fourth commit add integration tests of `Driver3SchemaBuilder`. Note that those integration tests do not check the handling of inconsistency between schemas.

The fifth commit documents an assumption I made that once a query is prepared, it will always return the data with the same schema (of course it might fail, but it won't return the data with another schema, even compatible one). This assumption is tested in integration tests. Protocol version V4 is explicitly set, as protocol V5 has different behavior (but it is not currently implemented by Scylla). 

This PR does not fully fix schema change handling as re-preparing the window queries is also needed.

Refs #45.
Refs scylladb/scylla#7824.

kbr- commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+         * Returns the column type for a column name in the base table,
+         * as of the time of CDC schema.
+         * <p>
+         * Proof of correctness under potential base table schema
```
But what is the hypothesis being proven?

haaawk replied:
That the class handles selected schema inconsistencies correctly, I presume.

avelanarius replied:
In the amended commit changed it to "Correctness of this implementation under possible inconsistencies with the base table schema" and "Correctness under possible base table schema inconsistencies" (term inconsistency is defined "near")

kbr- commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+         *         <td>The column data type change does not change
+         *             the column set (column names, whether the columns
+         *             are in the primary key). Therefore, the schemas
+         *             are consistent with regard to column type.</td>
```
I don't understand this one.
What does it mean that schemas are "consistent" when the column data type has changed?
See my previous comment - it's not clear what is actually being proven here (what the hypothesis is).

kbr- replied:
Ahh, I think I understand. The definition of cosistency that you're using is as follows:
Let `s` be CDC log schema built by this instance of `Driver3SchemaBuilder`. Let `b(s)` be the corresponding base table schema. (note that we may not have access to `b(s)` - we may have a different version of the base table schema, older or newer). Then the column kind returned by this function is _consistent_ if it is the same as the column kind in `b(s)`.
I think this definition (or a variation of it - the wording may be different) should be precisely stated before you start proving anything.

Still, I don't know if it makes sense to talk about "consistency of schema". It could, if we extend the above definition from column kind to schemas, e.g.: "let `b*(s)` be the schema described by this instance of `BaseTableSchema`. We say that `b*(s)` is consistent if it is equal to `b(s)`". Where "equal" means that getting the column kinds, column types etc. through the accessor functions returns the same values as it would if we were dealing directly with `b(s)`. (observational equality)

What is **really** confusing in this entire proof and basically the entire file is the overloaded "column type". Just read this: "Column data type changed. ... Therefore the schemas are consistent with regard to column type". What the heck. But the truth is it should say: "Column data type changed. ... Therefore the schemas are consistent with regard to column **kind**" - and now it makes sense.

avelanarius replied:
Added a definition of inconsistency in the amended commit. By inconsistency, I mostly mean a situation where `ALTER TABLE` happened and one schema (base table or CDC log schema) is old (before `ALTER TABLE`) and the other one is new (after `ALTER TABLE`).

kbr- commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+         * @param cdcColumnDefinitions the column definitions for the CDC log table.
+         * @return the column type in the base table.
+         */
+        public ChangeSchema.ColumnType getColumnType(String columnName, ColumnDefinitions cdcColumnDefinitions) {
```
Could you add a new commit before this one that changes `ColumnType` to `ColumnKind` (and then this function would be called getColumnKind)? Using ColumnType is reaaally confusing.

kbr- replied:
Even better: getBaseColumnKind?

avelanarius replied:
In the amended commit changed the naming to ColumnKind in the entire project (and all comments, etc.). This method is now called `getColumnKind`. I did not find it necessary to name it `getBaseColumnKind`, because a class it is in (`BaseTableSchema`) already implies this.

kbr- commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+        }
+
+        /**
+         * Returns the column type for a column name in the base table,
```
Not column kind?

kbr- reviewed (commented):
The overloaded use of "column type" (when we should be talking about "column kind") makes the code really hard to understand and review.

I've sent a batch of comments, but I think that before review proceeds further, it would be best to fix this misnomer.

avelanarius commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+                // was a LIST<X> in the base table.
+                if (cdcDataType.getCqlType() == ChangeSchema.CqlType.MAP
+                        && cdcDataType.getTypeArguments().get(0).getCqlType() == ChangeSchema.CqlType.TIMEUUID
+                        && baseTableSchema.isList(cdcColumn.getName())) {
```
The precondition to `isList` (and this precondition is already documented!) is that we know from the CDC schema that this column is a `list<X>` or `map<timeuuid, X>` in the base table schema (as of CDC log schema time).

`isList` throws inconsistency exception in a case it encounters a different type than `list<X>` or `map<timeuuid, X>`. This is for a case when someone changes a type from `list<X>` to `blob`, the CDC schema is old (`frozen<map<timeuuid, X>>`) and base table schema is fresh (`blob`). From such information you cannot determine if the type was `list<X>` or `map<timeuuid, X>` at the time of CDC schema.

If you'd remove the first and second conjuncts, it would throw an exception for each non-list, non-map-timeuuid collection type. Of course, you could move this logic (first and second conjuncts) from this place into `isList` - maybe a valid suggestion.

> Is it even possible to change non-frozen collection columns' types?

Yes, to `blob` (although I don't see such type change very useful...)

haaawk commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+ * must both be fresh.
+ * <p>
+ * For example, in case of <code>ALTER TABLE</code>,
+ * base table schema containing a schema before
+ * that operation and CDC log schema containing a schema
+ * after that operation, this builder will
```
How is it possible for CDC log schema to be after ALTER and base table schema to be before ALTER? I can imagine base table schema being after ALTER and CDC log being before but can't think of a way to achieve the opposite.

avelanarius replied:
Did some more real-life testing. Curiously enough, I am able to reproduce CDC log schema after ALTER and base table schema before ALTER, but not the other way around.

Code to reproduce:
```java
package com.datastax.driver.core;

import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.Session;

public class Testing {
    public static void main(String[] args) {
        Cluster c = Cluster.builder().addContactPoint("127.0.0.2").build();
        Session s = c.connect();

        while (true) {
            String baseTableType = c.getMetadata().getKeyspace("ks").getTable("t").getColumn("v").getType().getName().toString();
            String cdcTableType = c.getMetadata().getKeyspace("ks").getTable("t_scylla_cdc_log").getColumn("v").getType().getName().toString();
            if (!baseTableType.equals(cdcTableType)) {
                System.out.println("Base table is: " + baseTableType);
                System.out.println("CDC table is: " + cdcTableType);
                s.close();
                c.close();
                return;
            }
        }
    }
}
```

Before running: `CREATE TABLE ks.t(pk int, ck int, v int, PRIMARY KEY(pk, ck)) WITH cdc = {'enabled': true};`
While running: `ALTER TABLE ks.t ALTER v TYPE blob;`
Quite frequently you will see:
```
Base table is: int
CDC table is: blob
```
(CDC after ALTER, base before ALTER)

Right now I am having a difficulty reproducing the other way around. Since this is such a tricky thing, which could change with innocuous change in driver/Scylla, I think it is better to prepare for both cases.

avelanarius replied:
However, in actual code (this PR), the situation is a bit different (chronologically):

1. We prepare the statement on CDC log table.
2. We get the schema from the PreparedStatement on CDC log (first row returned by PreparedStatement).
3. We get the schema of base table from `Cluster`.

If `ALTER` happens before 1, CDC log schema will be fresh, but base table schema could be stale (generally there is 
at least ~1s delay in schema in `Cluster`). If `ALTER` happens after 1, CDC schema will be old, but maybe there was enough time for `Cluster` to acknowledge new schema, so the base table schema will be fresh.

I will consider updating a comment in the source code with this explanation above.

haaawk replied:
I see.

haaawk commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+                    String.format("Keyspace '%s' is missing from the cluster metadata.", keyspace));
+        }
+
+        TableMetadata tableMetadata = keyspaceMetadata.getTable(baseTableName);
```
I think it's not desirable that withRow has to be called before withClusterMetadata can be called. Otherwise baseTableName will be null here. Usually builders allow to call all the methods in random order. If withRow is required then let's take row in a constructor of the builder to make sure that baseTableName is always set when other methods are called.

avelanarius replied:
In the amended commit changed it to a factory. There is now only a single static `getChangeSchema` method.

haaawk commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3WorkerCQL.java:
```diff
@@ -127,9 +127,15 @@ public void onFailure(Throwable t) {
-                            .withRow(row).build();
+                    try {
+                        schema = new Driver3SchemaBuilder()
+                                .withRow(row)
+                                .withClusterMetadata(session.getCluster().getMetadata())
```
Here's an example where you had to swap the order of withRow and withClusterMetadata. This shouldn't be required with builder class. Otherwise we could just have a factory method that takes all necessary arguments.

haaawk commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+         *             are consistent with regard to column type.</td>
+         *     </tr>
+         *     <tr>
+         *         <td>CDC log schema newer than base schema</td>
```
How is this possible for CDC log schema to be newer than base table schema?

avelanarius replied:
Duplicate of previous comment.

haaawk commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+    private ChangeSchema.DataType translateCdcColumnDataType(DataType driverType) {
+        // This method only translates driverType
+        // from CDC table columns. All CDC table columns (with non-native types)
+        // are frozen, even though the driverType can lack this information
+        // (when driverType is from a PreparedStatement metadata).
```
Could you explain a bit more what's happening with the prepared statements that's special?

avelanarius replied:
In this specific line: We get the CDC log schema from `cdcRow.getColumnDefinitions()`. This object does not have information about whether the types are frozen or not (so they look like non-frozen). However, if you used the metadata from `cluster.getMetadata()`, it'd contain the information about whether the types are frozen or not.

haaawk commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -178,40 +531,31 @@ private void generatePrimaryKeyColumns() {
                         .collect(Collectors.collectingAndThen(Collectors.toList(), ImmutableList::copyOf));
                 Preconditions.checkArgument(!typeArguments.isEmpty());
-                return new ChangeSchema.DataType(ChangeSchema.CqlType.TUPLE, typeArguments, driverType.isFrozen());
+                return new ChangeSchema.DataType(ChangeSchema.CqlType.TUPLE, typeArguments, true);
```
A localised comment explaining that all CDC columns are frozen would be nice here.

avelanarius commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+ * <p>
+ * The implementation does not require that
+ * the base table schema and CDC log schema
+ * must both be fresh.
```
Fixed in the amended commit.

avelanarius commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
-    private Set<String> baseTablePartitionKeyColumnNames;
-    private Set<String> baseTableClusteringKeyColumnNames;
 
     public Driver3SchemaBuilder withRow(Row row) {
```
In the amended commit, this class is now a factory (not a builder), so there is a single `getChangeSchema` method with no call-method-X-before preconditions.

avelanarius commented on scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java:
```diff
@@ -12,125 +13,477 @@
+        if (!cdcDataType.isNative()) {
+            // In the CDC table we have a frozen non-native type (frozen collection,
+            // frozen UDT or frozen tuple). That means that in the base table there
+            // is a same type - but it could be frozen or non-frozen.
```
In the amended commit changed this comment to:
```java
// In the CDC table we have a frozen non-native type (frozen collection,
// frozen UDT or frozen tuple). However, both non-frozen and frozen
// non-native types in the base table map to a frozen non-native
// type in the CDC log table.
// We use the existence of cdc$deleted_elements_ column as a proxy
// for the information whether the column was frozen or non-frozen
// in the base table.
```

avelanarius commented:
v2:
- Renamed `ColumnType` to `ColumnKind` (in the entire project). The old `ColumnType` is not deleted, but deprecated (will probably delete it "next year", but I want to limit the number of breaking changes).
- Fixed some comments (grammatical issues, better "inconsistency" definition)
- Made `Driver3SchemaBuilder` a factory, instead of a builder (which required a specific ordering of called methods)
- Added JavaDocs to `getChangeSchema`

haaawk reviewed (approved):
LGTM
